homeassistant/components/mqtt/config_flow.py

```diff
@@ -337,38 +452,149 @@ def validate_field(
+        # Omit empty lists that are not allowed to be empty
+        if isinstance(value, dict):
+            merged_user_input.update(value)
+        else:
+            merged_user_input[key] = value
+
+    for field, value in merged_user_input.items():
         validator = data_schema_fields[field].validator
         try:
             validator(value)
         except (ValueError, vol.Invalid):
             errors[field] = data_schema_fields[field].error or "invalid_input"
 
+    if config_validator is not None:
+        if TYPE_CHECKING:
+            assert component_data is not None
+
+        config_validator(
+            calculate_merged_config(
+                merged_user_input, data_schema_fields, component_data
+            ),
+            errors,
+            reset_fields,
+        )
+
+    for field in reset_fields:
+        if component_data and field in component_data:
+            del component_data[field]
+
+    return merged_user_input
+
 
 @callback
 def data_schema_from_fields(
     data_schema_fields: dict[str, PlatformField],
     reconfig: bool,
+    component: dict[str, Any] | None = None,
+    user_input: dict[str, Any] | None = None,
 ) -> vol.Schema:
-    """Generate data schema from platform fields."""
-    return vol.Schema(
-        {
+    """Generate custom data schema from platform fields."""
+    user_data = deepcopy(component)
+    if user_data is not None and user_input is not None:
+        user_data |= user_input
+    sections: dict[str | None, None] = {
+        field_details.section: None for field_details in data_schema_fields.values()
+    }
+    data_schema: dict[Any, Any] = {}
+    all_data_element_options: set[Any] = set()
+    no_reconfig_options: set[Any] = set()
+    for schema_section in sections:
+        data_schema_element = {
             vol.Required(field_name, default=field_details.default)
             if field_details.required
             else vol.Optional(
                 field_name, default=field_details.default
-            ): field_details.selector
+            ): field_details.selector(user_data)  # type: ignore[operator]
+            if field_details.custom_filtering
+            else field_details.selector
             for field_name, field_details in data_schema_fields.items()
-            if not field_details.exclude_from_reconfig or not reconfig
+            if field_details.section == schema_section
+            and (not field_details.exclude_from_reconfig or not reconfig)
+            and _check_conditions(field_details, user_data)
         }
-    )
+        data_element_options = set(data_schema_element)
+        all_data_element_options |= data_element_options
+        no_reconfig_options |= {
+            field_name
+            for field_name, field_details in data_schema_fields.items()
+            if field_details.section == schema_section
+            and field_details.exclude_from_reconfig
+        }
+        if not data_element_options:
```

Comment:
These lines have no test coverage yet. They will be effective when fields in sections are filtered out based on previous data in the flow.